### PR TITLE
kv: support `ResultsFilteredByACLs` flag/header

### PR DESCRIPTION
**Pull request by boxofrad**
Adds support for the `ResultsFilteredByACLs` flag, and corresponding `X-Consul-Results-Filtered-By-ACLs` HTTP header (introduced in #11569) to the KV `List` and `ListKeys` endpoints.

### CODE REVIEW

File: agent/consul/kvs_endpoint.go
Comment (FFMMM) [inline agent/consul/kvs_endpoint.go:210]: [nit] hm, could be a style preference but curious if we could encapsulate the logic here in a func that can be used for `List` and ListKeys`?

File: agent/consul/kvs_endpoint_test.go
Comment (FFMMM) [inline agent/consul/kvs_endpoint_test.go:486]: could we add a test here (or somewhere else) to test that `ResultsFilterdByACLs` is set for the `List` endpoint as expected?

File: agent/consul/kvs_endpoint_test.go
Comment (FFMMM) [inline agent/consul/kvs_endpoint_test.go:292]: I think this suffices as a test but I was curious if we could actually test for the presence of the header somehow as well.

But this may actually be covered, generally, thru https://github.com/hashicorp/consul/pull/11569 ; need to double check

File: agent/consul/kvs_endpoint.go
Comment (boxofrad) [reply agent/consul/kvs_endpoint.go:210]: Maybe, yeah! Were you thinking of something like `k.filterDirEnt(&reply)` that does the filtering and sets the flag?

File: agent/consul/kvs_endpoint_test.go
Comment (boxofrad) [reply agent/consul/kvs_endpoint_test.go:486]: I think this is covered here:

https://github.com/hashicorp/consul/blob/b5f81dd0c0f8361e986dbb86fde96fe5644f50da/agent/consul/kvs_endpoint_test.go#L481-L483

File: agent/consul/kvs_endpoint_test.go
Comment (boxofrad) [reply agent/consul/kvs_endpoint_test.go:292]: Ah yeah, it is covered more generally by the `setMeta` tests in that PR, but we certainly could have better coverage by testing it on the individual HTTP handlers themselves.

On balance, I decided that, aside from the HTTP-only endpoints (in #11610), testing at the RPC level is an acceptable trade-off of value vs maintainance burden.

What do you think? 🙇🏻‍♂️

**Review by FFMMM — APPROVED** [review]
🚀 this LGTM! Up to you if you want to tackle that nit!

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/515836.

**Pull request merged**
